engine/utils/Defer.h

```diff
@@ -0,0 +1,54 @@
+//
+// Copyright (c) 2023 Jonathan Moallem (@J-Mo63) & Aryeh Zinn (@Raelr)
+//
+// This code is released under an unmodified zlib license.
+// For conditions of distribution and use, please see:
+//     https://opensource.org/licenses/Zlib
+//
+
+#ifndef SIEGE_ENGINE_DEFER_H
+#define SIEGE_ENGINE_DEFER_H
+
+#include <functional>
+
+#include "Macros.h"
+
+namespace Siege
+{
+
+/**
+ * A basic deferred call object implementation that executes its provided callable on scope exit.
+ */
+struct ScopedDefer
+{
+    /**
+     * Explicitly deleted zero-param constructor
+     */
+    ScopedDefer() = delete;
+
+    /**
+     * Constructor, specifying owned callable
+     * @param callable - the void-returning, parameterless callable to execute
+     */
+    explicit ScopedDefer(const std::function<void()>& callable) : callable(callable) {}
```

Comment:
I'm wondering if there's a way we could do this without using `std::function`. I wonder if a function pointer could achieve the same thing here? Not the end of the world if we can't, just a general thought.

Reply:
`std::function` is more or less a function pointer - just a strongly typed one that nicely handles lambda.